### PR TITLE
feat: add event templating

### PR DESCRIPTION
**Pull request by jorgeantonio21**
Description
---
Adds events to the Tari Runtime. It also corrects a minor bug on the runtime tracker `finalize` method, that would ignore logging.

Motivation and Context
---

How Has This Been Tested?
---
Add test `basic_emit_event` on the engine tests.

What process can a PR reviewer use to test or verify this change?
---

Breaking Changes
---

- [x] None
- [ ] Requires data directory to be deleted
- [ ] Other - Please specify

### CODE REVIEW

File: applications/tari_dan_app_grpc/proto/transaction.proto [hunk 1]
```diff
@@ -46,6 +46,7 @@ message Instruction {
     PUT_OUTPUT_IN_WORKSPACE = 2;
     EMIT_LOG = 3;
     CLAIM_BURN = 4;
+    EMIT_EVENT = 5;
```
Comment (mrnaveira) [inline applications/tari_dan_app_grpc/proto/transaction.proto]: I am not convinced that we need a new Instruction type for events, as I understand events should be emitted only by components and not by the user. That would also mean that we don't need the extra fields in transaction/instruction types.

File: dan_layer/engine/src/runtime/impl.rs [hunk 2]
```diff
@@ -167,6 +175,7 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> RuntimeInte
             LogLevel::Warn => log::Level::Warn,
             LogLevel::Info => log::Level::Info,
             LogLevel::Debug => log::Level::Debug,
+            LogLevel::Store => todo!("Implement me later"),
```
Comment (mrnaveira) [inline dan_layer/engine/src/runtime/impl.rs]: As we are using a separated event function, we probably need to remove the `Store` variant here

File: dan_layer/engine/src/runtime/impl.rs [hunk 3]
```diff
@@ -821,11 +830,13 @@ impl<TTemplateProvider: TemplateProvider<Template = LoadedTemplate>> RuntimeInte
         let substates_to_persist = self.tracker.take_substates_to_persist();
         self.invoke_modules_on_before_finalize(&substates_to_persist)?;
 
-        let (result, fee_receipt) = self.tracker.finalize(substates_to_persist)?;
+        let events = self.tracker.take_events();
```
Comment (mrnaveira) [inline dan_layer/engine/src/runtime/impl.rs]: At this point we may need a new struct to define the tracker finalization, that will include `result`, `fee_receipt` and the `events` as fields.

File: dan_layer/engine_types/src/events.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct Event {
+    pub timestamp: u64,
+    pub message: String,
+}
+
+impl Event {
+    pub fn new(message: String) -> Self {
+        let now = SystemTime::now()
```
Comment (mrnaveira) [inline dan_layer/engine_types/src/events.rs]: I am pretty sure that this way of getting the timestamp will not work in a real network (or even in a integration test), as each VN in the committee will get a different value thus the result of the transaction will be different as well.

We need a timestamp value that is calculated exactly the same in all VNs in the committee. That is probably non-trivial, so for now I would leave timestamps outside of the scope.

**Review by mrnaveira — DISMISSED** [review]
The approach is good in general, but some changes are necessary IMO

File: dan_layer/engine/src/runtime/impl.rs (hunk 2 above)
Comment (jorgeantonio21) [reply dan_layer/engine/src/runtime/impl.rs]: Great catch, remove it

File: dan_layer/engine_types/src/events.rs (hunk 4 above)
Comment (jorgeantonio21) [reply dan_layer/engine_types/src/events.rs]: Very good observation, agreed ! Removed the field

File: applications/tari_dan_app_grpc/proto/transaction.proto (hunk 1 above)
Comment (jorgeantonio21) [reply applications/tari_dan_app_grpc/proto/transaction.proto]: Removed it

File: dan_layer/transaction_manifest/src/parser.rs [hunk 5]
```diff
@@ -324,6 +324,7 @@ fn macro_call(mac: &Ident, tokens: TokenStream) -> Result<ManifestIntent, syn::E
             level: LogLevel::Error,
             message: parse2::<LitStr>(tokens)?.value(),
         })),
+        "store" => todo!("Implement me later"),
```
Comment (sdbondi) [inline dan_layer/transaction_manifest/src/parser.rs]: Is this needed? What is the intention?

**Review by sdbondi — DISMISSED** [review]
Looks like a good start

File: dan_layer/transaction_manifest/src/parser.rs (hunk 5 above)
Comment (jorgeantonio21) [reply dan_layer/transaction_manifest/src/parser.rs]: It was supposed to be removed, but missed it. Thanks for catching it :)

**Review by sdbondi — APPROVED** [review]

**Pull request merged**